File: src/render/theme.py
"""Palette, type ramp and metrics shared by the tray icons and the panel.

Two color ideas live here. A brand accent marks the parts of the interface that
belong to the product, the mark, the refresh control and the spinner; it comes
from the provider, so the panel is orange for Claude and green for ChatGPT. The
usage ramp is a separate scale that runs from green to red with the percentage,
so a bar says how close to its limit it is by color alone, without anyone having
to read the number.

The rest follows Apple's dark system palette and type scale: a near black
background, label colors expressed as white at decreasing prominence, and a
small set of named text styles built from size and weight together.
"""

from __future__ import annotations

from pathlib import Path

from ..validation import require_member, require_number_in_range, require_positive_int

# Stops of the usage ramp as (percentage, (red, green, blue)). A value between
# two stops is mixed from them, so the bar shifts continuously rather than
# jumping at a threshold.
USAGE_RAMP = (
	(0.0, (48, 209, 88)),
	(50.0, (227, 193, 59)),
	(75.0, (232, 145, 45)),
	(100.0, (255, 69, 58)),
)

# Opacity of a bar track, which is the bar color laid over the background.
TRACK_OPACITY = 0.22

ICON_LABEL = (255, 255, 255, 255)
TRANSPARENT = (0, 0, 0, 0)

SURFACE_BASE = "#1c1c1e"

# Label colors, brightest first, matching label, secondary and tertiary.
LABEL_PRIMARY = "#ffffff"
LABEL_SECONDARY = "#a1a1a6"
LABEL_TERTIARY = "#6e6e73"

UI_FONT_FAMILY = "Segoe UI"

# Size and weight together, in the order of Apple's text styles. Sizes are in
# points, so Windows scales them with the display and the user text size.
TEXT_STYLES = {
	"title": (12, "bold"),
	"section": (10, "bold"),
	"row": (10, "normal"),
	"row_value": (10, "normal"),
	"caption": (8, "normal"),
	"caption_strong": (8, "bold"),
}

FONT_CANDIDATES = {
	"bold": (
		"C:/Windows/Fonts/segoeuib.ttf",
		"C:/Windows/Fonts/seguisb.ttf",
		"C:/Windows/Fonts/arialbd.ttf",
		"C:/Windows/Fonts/DejaVuSans-Bold.ttf",
	),
	"normal": (
		"C:/Windows/Fonts/segoeui.ttf",
		"C:/Windows/Fonts/arial.ttf",
		"C:/Windows/Fonts/DejaVuSans.ttf",
	),
}


def usage_rgb(percent: float) -> tuple[int, int, int]:
	"""Return the ramp color for a usage percentage.

	Args:
		percent: Share of a window already used, 0 to 100. float.

	Returns:
		tuple[int, int, int]: The red, green and blue components, each 0 to 255.
	"""
	require_number_in_range(percent, 0.0, 100.0, "percent")
	previous_stop, previous_color = USAGE_RAMP[0]
	for stop, color in USAGE_RAMP:
		if percent <= stop:
			span = stop - previous_stop
			ratio = 0.0 if span <= 0.0 else (percent - previous_stop) / span
			return tuple(int(round(low + (high - low) * ratio)) for low, high in zip(previous_color, color))
		previous_stop, previous_color = stop, color
	return USAGE_RAMP[-1][1]


def usage_color(percent: float) -> tuple[int, int, int, int]:
	"""Return the ramp color for a usage percentage as an opaque RGBA tuple.

	Args:
		percent: Share of a window already used, 0 to 100. float.

	Returns:
		tuple[int, int, int, int]: The RGBA color to fill an icon bar with.
	"""
	return usage_rgb(percent) + (255,)


def usage_hex(percent: float) -> str:
	"""Return the ramp color for a usage percentage as a hexadecimal string.

	Tkinter accepts hexadecimal strings rather than RGBA tuples, so the ramp is
	exposed in both forms.

	Args:
		percent: Share of a window already used, 0 to 100. float.

	Returns:
		str: The color as a "#rrggbb" string.
	"""
	return "#{:02x}{:02x}{:02x}".format(*usage_rgb(percent))


def text_style(name: str) -> tuple[str, int, str]:
	"""Return a named text style as a Tk font specification.

	Args:
		name: A key of TEXT_STYLES, such as "title" or "caption". str.

	Returns:
		tuple[str, int, str]: The family, size in points and weight, ready to
		pass as a Tk font option.
	"""
	require_member(name, frozenset(TEXT_STYLES), "name")
	size, weight = TEXT_STYLES[name]
	return (UI_FONT_FAMILY, size, weight)


def mix_hex(background: str, foreground: str, amount: float) -> str:
	"""Return the color that results from laying one color over another.

	Tkinter has no alpha channel, so translucent tracks and pressed states are
	flattened against the surface they sit on.

	Args:
		background: The lower color as a "#rrggbb" string. str.
		foreground: The upper color as a "#rrggbb" string. str.
		amount: Opacity of the upper color, 0 to 1. float.

	Returns:
		str: The blended color as a "#rrggbb" string.
	"""
	require_number_in_range(amount, 0.0, 1.0, "amount")
	lower = tuple(int(background.lstrip("#")[index : index + 2], 16) for index in (0, 2, 4))
	upper = tuple(int(foreground.lstrip("#")[index : index + 2], 16) for index in (0, 2, 4))
	blended = tuple(int(round(low + (high - low) * amount)) for low, high in zip(lower, upper))
	return "#{:02x}{:02x}{:02x}".format(*blended)


def track_hex(percent: float) -> str:
	"""Return the track color for a bar, which is the bar color dimmed.

	Args:
		percent: Share of the window already used, 0 to 100. float.

	Returns:
		str: The color as a "#rrggbb" string.
	"""
	return mix_hex(SURFACE_BASE, usage_hex(percent), TRACK_OPACITY)


def load_font(size: int, weight: str = "bold"):
	"""Return a TrueType font of the platform family at the requested size.

	Pillow needs a face on disk where Tk takes a family name, so the same type
	the panel sets in Tk is loaded here by file.

	Falls back to the Pillow bitmap font when no candidate face is installed, so
	rendering never fails on a machine with an unusual font set.

	Args:
		size: Font size in pixels. int, greater than 0.
		weight: Either "normal" or "bold". str.

	Returns:
		PIL.ImageFont.ImageFont: A font object usable with ImageDraw.text.
	"""
	require_positive_int(size, "size")
	require_member(weight, frozenset(FONT_CANDIDATES), "weight")
	from PIL import ImageFont

	for candidate in FONT_CANDIDATES[weight]:
		if Path(candidate).exists():
			try:
				return ImageFont.truetype(candidate, size)
			except OSError:
				continue
	return ImageFont.load_default()

File: src/mac/FORL/Design/Theme.swift
//
//  Theme.swift
//  Color and type for every surface.
//
//  Almost nothing is a fixed value. Label colors, separators and backgrounds are
//  asked of the system by name, so they follow the appearance the user is in and
//  the increased contrast setting without this file having to notice either.
//
//  Two colors are the app's own. A brand accent marks what belongs to the
//  product, the mark and the dial, and comes from the provider. The usage ramp
//  runs from green to red with the percentage, so a bar says how close to its
//  limit it is by color alone, without anyone having to read the number.
//

import SwiftUI

/// Stops of the usage ramp. A value between two stops is mixed from them, so a
/// gauge shifts continuously rather than jumping at a threshold.
private let usageRamp: [(stop: Double, color: (Double, Double, Double))] = [
	(0, (48, 209, 88)),
	(50, (227, 193, 59)),
	(75, (232, 145, 45)),
	(100, (255, 69, 58)),
]

/// The measurements and colors every surface shares.
enum Theme {
	/// Opacity of a gauge track, which is the gauge color laid over whatever is
	/// behind it.
	static let trackOpacity: Double = 0.22

	/// Opacity of a rule between two groups.
	static let hairlineOpacity: Double = 0.09

	/// Thickness of a dial's ring as a share of its own diameter, which is what
	/// Apple draws a battery dial at: thin enough to read as a gauge rather than
	/// as a donut.
	static let ringThickness: Double = 0.085

	/// How far across a dial the ring reaches, as a share of its square.
	static let ringRadius: Double = 0.36

	/// How a gauge moves. It settles without overshoot, because a usage reading
	/// that bounced past its value would be reporting something that never
	/// happened.
	static let gaugeMotion: Animation = .spring(duration: 0.4, bounce: 0)

	/// Return the ramp color for a usage percentage.
	///
	/// - Parameter percent: Share of a window already used, 0 to 100.
	/// - Returns: The color a gauge at that percentage is filled with.
	static func usageColor(_ percent: Double) -> Color {
		let value = clampPercent(percent)
		var previous = usageRamp[0]
		for entry in usageRamp {
			if value <= entry.stop {
				let span = entry.stop - previous.stop
				let ratio = span <= 0 ? 0 : (value - previous.stop) / span
				return Color(
					red: mix(previous.color.0, entry.color.0, ratio) / 255,
					green: mix(previous.color.1, entry.color.1, ratio) / 255,
					blue: mix(previous.color.2, entry.color.2, ratio) / 255
				)
			}
			previous = entry
		}
		let last = usageRamp[usageRamp.count - 1].color
		return Color(red: last.0 / 255, green: last.1 / 255, blue: last.2 / 255)
	}

	/// Return the color of a gauge track.
	///
	/// A gauge with nothing to report takes a gray track. The ramp starts at
	/// green, and a green gauge would be claiming the limit is untouched at the
	/// moment the app does not yet know what it is.
	///
	/// - Parameter percent: Share of the window already used, or nil when no
	///   reading has arrived.
	/// - Returns: The color to fill the whole length of the gauge with.
	static func trackColor(_ percent: Double?) -> Color {
		guard let percent else {
			return Color.primary.opacity(0.18)
		}
		return usageColor(percent).opacity(trackOpacity)
	}

	/// Return one component mixed between two stops.
	///
	/// - Parameters:
	///   - low: The component at the lower stop.
	///   - high: The component at the upper stop.
	///   - ratio: How far between them, 0 to 1.
	/// - Returns: The mixed component.
	private static func mix(_ low: Double, _ high: Double, _ ratio: Double) -> Double {
		low + (high - low) * ratio
	}
}
